Force Cabal >= 1.24 dep when there's no custom-setup stanza.

23Skidoo opened the pull request:
Fixes #3199. 

To minimise the impact we should probably add default setup dependencies only to packages affected by #3199 (see @grayjay's [snippet](https://gist.github.com/grayjay/c5c2627d02303bd59a3b164055233bcb)). Note, however, that the setup wrapper will pick Cabal >= 1.24 anyway if it is installed, since it prefers new Cabal versions.

Note also that `cabal configure` will still fail (unlike `cabal install`), though the error message will be more helpful than it currently is:

```
Resolving dependencies...
Warning: solver failed to find a solution:
Could not resolve dependencies:
trying: test-t3199-0.1.0.0 (user goal)
next goal: test-t3199-setup.Cabal (dependency of test-t3199-0.1.0.0)
rejecting: test-t3199-setup.Cabal-1.22.5.0/installed-cfe... (conflict:
test-t3199 => test-t3199-setup.Cabal>=1.24)
Dependency tree exhaustively searched.
Trying configure anyway.
Configuring test-t3199-0.1.0.0...
```

/cc @dcoutts

23Skidoo commented:
I considered adding a test case, but it's a bit difficult given that you need to have a pre-1.24 version of Cabal installed for the bug to manifest.

23Skidoo commented:
GHC 8 failure is due to GHC 8 having an old snapshot of Cabal 1.24. Ugh, wish we could ignore it somehow.

ezyang commented:
In principle, cabal-install should know what version of Cabal it was built with, so it can demand that from the package database. But this won't work if you're distributing a statically linked Cabal binary, in that case you can't assume the Cabal 1.24 library is installed.

23Skidoo commented:
> But this won't work if you're distributing a statically linked Cabal binary

Yep. Some people also like to build `cabal-install` inside a sandbox and then copy the executable to `~/bin`.

ezyang commented:
I don't understand the approach this patch takes. Wasn't the original plan to only levy this implicit dependency if there was a `buildable` field?

ezyang commented:
I think we want a test, if only to tell if the patch actually fixes the problem.

I think the easiest way to do this is to make the integration test GHC < 8.0 only. In these cases, we can assume GHC was distributed with an older version of Cabal. This will be sufficient for Travis.

23Skidoo commented:
> I don't understand the approach this patch takes. Wasn't the original plan to only levy this implicit dependency if there was a `buildable` field?

Sure, I can do that, just wanted to get some comments first. Though I think that forcing custom setup scripts to use the same Cabal version that cabal-install was compiled with is a good idea in general (provided that the user can force a different version manually via `--use-cabal-version`).

ezyang commented on cabal-install/Distribution/Client/SetupWrapper.hs:
```diff
@@ -179,6 +181,13 @@ data SetupScriptOptions = SetupScriptOptions {
+    -- 'useDependencies'. This is turned on (in conjunction with
+    -- useDependenciesExclusive = False) when a default setup dependencies
+    -- stanza is added by cabal-install itself. See #3199.
+    noDefaultCabalDep :: Bool,
```
So, is useDependenciesExclusive == True, noDefaultCabalDep == True and impossible state?

23Skidoo replied:
> `useDependenciesExclusive == True`

In that case `noDefaultCabalDep` is ignored.

dcoutts replied:
I hope we can simplify the setup wrapper in future, the logic is getting really rather hairy. Fortunately the new-build code path only needs a fraction of the functionality, since it knows exactly what deps to use in advance, so doesn't need any of the guesswork that the setupWrapper currently does.

ezyang commented:
CC @dcoutts I know Duncan has stated we want the logic for compiling custom setups to be as simple as possible, so other tools (like Stack) can also handle them; I don't know how far the wrong way this goes.

ezyang commented:
Re the Travis failure, a minor version bump should do the trick, right?

23Skidoo commented:
Actually I think I can get rid of `noDefaultCabalDep`.

> Re the Travis failure, a minor version bump should do the trick, right?

Yes, I just didn't want the first 1.24 release to have a version number other than 1.24.0.0. Next time we create a release branch we should bump the version only just before the release.

dcoutts commented:
@23Skidoo Unfortunately the first patch would break the use case in new-build. Have a look at:

https://github.com/haskell/cabal/blob/1223a395151b14d7899e8a93477d7b793314e589/cabal-install/Distribution/Client/ProjectPlanning.hs#L1624-L1668

There we apply custom deps for two cases, build-type custom with no explicit deps, but also non-custom when we need to build an external setup anyway.

So rather than changing how `addDefaultSetupDependencies` works, instead all you need to do is to change the function you pass to it, so that like `defaultSetupDeps` in the new-build code, you only return non-[] in the cases you're interested in, which in this case is only build-type custom with no explicit deps.

dcoutts commented on Cabal/Distribution/PackageDescription.hs:
```diff
@@ -308,18 +308,22 @@ instance Text BuildType where
 data SetupBuildInfo = SetupBuildInfo {
-        setupDepends :: [Dependency]
+        setupDepends        :: [Dependency],
+        defaultSetupDepends :: Bool
```
Ah, cunning. I went to some lengths in the new-build code to do this without changing the underlying type, but yes it is rather easier if we can just stash the info here. I might switch the new-build code over to use that, rather than keeping an extra Set around.

dcoutts commented on cabal-install/Distribution/Client/Dependency.hs:
```diff
@@ -451,6 +452,10 @@ standardInstallPolicy
+  . addDefaultSetupDependencies
+      (const [Dependency (PackageName "Cabal")
+              (orLaterVersion $ Version [1,24] [])])
+
```
So make this function conditional on the build-type Custom, and then we don't need the previous patch to apply that conditionality within `addDefaultSetupDependencies` itself.

dcoutts commented:
So this fixes things for packages without custom setup deps, of course we still have the problem for packages that do specify an upper bound we would still have trouble communicating to them.

But it's ok as a quick hack before the release I'd say (modulo the fixes I've suggested above), but it'd be good to try and make it possible to use older Cabal versions, since we do want to be able to have packages explicitly specify versions they work with (especially since we do want to start breaking the Cabal lib API to let us make bigger improvemtns). So we may need a proper fix to this in master at some point, and perhaps also consider making releases off of older branches that work with newer compiler versions.

hvr commented:
> of course we still have the problem for packages that do specify an upper bound we would still have trouble communicating to them

That's no different than a package specifying an upper bound on e.g. `base`. The solver ought to avoid such install-plans. Just as with ordinary upper bounds, we can introduce a `--setup-allow-newer=Cabal` flag...

> perhaps also consider making releases off of older branches that work with newer compiler versions.

...but then cabal would also need a way to know that, for instance, the preinstalled `Cabal-1.16.0` version that comes with GHC 7.6.3 is incompatible, but instead `Cabal-1.16.0.1` is compatible. It's a much simpler model imho to assume that there's a cut-off version of Cabal lib versions, e.g. `Cabal >= 1.18` which the current `cabal-install` is able to speak to.

@23Skidoo 

> Next time we create a release branch we should bump the version only just before the release.

that would cause problems for GHC dev (we request a bump-early-in-git policy especially for major ver bumps from bundled libraries, see https://ghc.haskell.org/trac/ghc/wiki/Commentary/Libraries/EagerVersionBump) and mask problems during RCs and/or cause annoying upper-bound-change-waves if last-minute upper bound changes occur :-/

23Skidoo commented:
@hvr
> Just as with ordinary upper bounds, we can introduce a `--setup-allow-newer=Cabal` flag...

`--allow-newer` already works for setup dependencies.

23Skidoo commented:
PR updated:

1. Removed `noDefaultCabalDep`.
1. Removed the first patch.
1. Limited the scope only to packages potentially affected by #3199.
1. Bumped Cabal version.
1. Added a test case.

@hvr 
> that would cause problems for GHC dev (we request a bump-early-in-git policy especially for major ver bumps from bundled libraries

Our versioning scheme (odd numbers == development, even == official releases) worked fine for GHC snapshots so far, I just want to extend it for release branches. So e.g. the 1.24 branch would continue to have the 1.23.x.y version right until the release. If reverse dependencies will specify the Cabal dep as `>= 1.23.x.y && < 1.25` then there should be no problems, I think.

The invariant I want to enforce is "version of Cabal in the repo" > "version of Cabal that came with GHC we're compiling with".

ezyang commented on cabal-install/Distribution/Client/Dependency.hs:
```diff
@@ -449,12 +450,62 @@ standardInstallPolicy
+          alwaysTrue (PD.Lit True) = True
+          alwaysTrue _          = False
+
+          -- copied from D.PackageDescription.Configuration.extractCondition
```
Blugh, copypasta!!

ezyang commented on cabal-install/tests/IntegrationTests/regression/t3199/ChangeLog.md:
```diff
@@ -0,0 +1,5 @@
+# Revision history for logop
```
As a matter of personal preference, I always remove this (and the LICENSE)

ezyang commented on cabal-install/tests/IntegrationTests/regression/t3199/test-3199.cabal:
```diff
@@ -0,0 +1,29 @@
+name:                test-t3199
+version:             0.1.0.0
+license:             BSD3
+license-file:        LICENSE
```
And then drop the license-field and extra-source-files fields

23Skidoo replied:
Fixed.

ezyang commented on cabal-install/tests/IntegrationTests/regression/t3199.sh:
```diff
@@ -0,0 +1,8 @@
+
+GHC_VERSION=`ghc --numeric-version`
+
+if [[ $GHC_VERSION < "8" ]]; then
```
I don't think this will have the intended semantics when GHC 10 is released.

ezyang commented on cabal-install/tests/IntegrationTests/regression/common.sh:
```diff
@@ -0,0 +1,9 @@
+# Helper to run Cabal
+cabal() {
+    "$CABAL" $CABAL_ARGS "$@"
```
These files are no longer needed unless you need to override the default common.sh (which it does not seem to be the case.)

23Skidoo replied:
> These files are no longer needed 

Did that change make it into 1.24?

ezyang commented on cabal-install/Distribution/Client/SetupWrapper.hs:
```diff
@@ -604,16 +605,22 @@ externalSetupMethod verbosity options pkg bt mkargs = do
           --
           -- Both of these options should be enabled for packages that have
           -- opted-in and declared a custom-settup stanza.
           --
```
Ugh, this is annoying. Under what circumstances do we have `not useDependenciesExclusive` but we DO have a Cabal? (I guess something related to the default setup deps?)

23Skidoo replied:
When we specify Cabal as a default setup dependency we want to prevent setup wrapper from adding another Cabal dependency. But we don't want it to set `-hide-all-packages`, since legacy setup scripts sometimes import stuff from libraries other than `base`/`Cabal`.

ezyang commented on cabal-install/Distribution/Client/ProjectPlanning.hs:
```diff
@@ -1575,14 +1575,15 @@ packageSetupScriptStylePreSolver pkg
 -- See 'rememberImplicitSetupDeps'.
 --
-defaultSetupDeps :: Platform -> PD.PackageDescription -> [Dependency]
+defaultSetupDeps :: Platform -> PD.PackageDescription -> Maybe [Dependency]
```
Maybe we don't need this change anymore?

23Skidoo replied:
May be. I wanted to minimise the impact of my changes.

ezyang commented on cabal-install/Distribution/Client/Dependency.hs:
```diff
@@ -449,12 +450,62 @@ standardInstallPolicy
+          noDepConditions = extractConditions (null . PD.targetBuildDepends)
+          extractConditions f =
+            concat [
+              maybeToList $ PD.extractCondition (f . PD.libBuildInfo)
```
It's a bit annoying if we don't already have some sort of fold function to do this logic...

23Skidoo commented:
The test case could be improved: right now we always install Cabal 1.24 before running the `cabal-install` test suite on Travis, so the failure can't manifest.

ezyang commented:
What if we tell Cabal to not use the user database for these tests?

23Skidoo commented:
Something like that should work.

23Skidoo commented:
Will fix the remaining issues tomorrow.

23Skidoo commented:
Updated the test case.

23Skidoo commented:
Moved copy-pasted utilities into the Cabal library, improved the test case. I think that this is now ready to go in, will merge once the build bots are green.

grayjay commented on Cabal/Distribution/PackageDescription.hs:
```diff
@@ -1193,11 +1199,30 @@ data Condition c = Var c
+cOr  x           (Lit False) = x
+cOr  c           (CNot d)
+  | c == d                   = Lit True
+cOr  x           y           = COr x y
```
I forgot to mention that I added a case to my snippet:

    cor  (CNot c)    d
      | c == d                   = Lit True

Without it, the function testing whether #3199 applies doesn't ignore this use of buildable:
```
executable exe
  if !flag(build-exe)
    buildable: False
  else
    build-depends: package
  [...]
```

hvr replied:
indeed, it's kinda odd that symmetry was broken for the "x or (not x)" case...

23Skidoo replied:
@grayjay Updated.
